### PR TITLE
[FLINK-4451] [rpc] Throw RpcConnectionException when rpc endpoint is not reachable

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR introduces a RpcConnectionException which is thrown if the rpc endpoint
is not reachable when calling RpcService.connect.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/registration/RetryingRegistration.java [hunk 1]
```diff
@@ -197,7 +197,7 @@ public void onSuccess(Gateway result) {
 				@Override
 				public void onFailure(Throwable failure) {
 					if (!isCanceled()) {
-						log.warn("Could not resolve {} address {}, retrying...", targetName, targetAddress);
+						log.warn("Could not resolve {} address {}, retrying...", targetName, targetAddress, failure);
```
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/registration/RetryingRegistration.java]: I think you cannot mix argument markers and exception logging.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActorTest.java [hunk 2]
```diff
@@ -73,6 +75,22 @@ public void testAddressResolution() throws Exception {
...
 
 	/**
+	 * Tests that a {@link RpcConnectionException} is thrown if the rpc endpoint cannot be connected to.
+	 */
+	@Test
+	public void testFailingAddressResolution() throws Exception {
+		Future<DummyRpcGateway> futureRpcGateway = akkaRpcService.connect("foobar", DummyRpcGateway.class);
+
+		try {
+			DummyRpcGateway gateway = Await.result(futureRpcGateway, timeout.duration());
```
Comment (StephanEwen) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActorTest.java]: Will this fail immediately, or only after a timeout?

**Comment by StephanEwen** [comment]
The idea makes sense. How does the system determine that the RPC endpoint is not reachable? By the fact that the initial "identify" message times out?

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActorTest.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/akka/AkkaRpcActorTest.java]: This should fail as soon as the future is completed (in this case failed with the exception). The exception is thrown as soon as Akka returns an empty `ActorIdentity` message.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/registration/RetryingRegistration.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/registration/RetryingRegistration.java]: I think you can according to http://stackoverflow.com/a/6374166/4815083 and I also just tested it.

**Comment by tillrohrmann** [comment]
Akka will return an empty `ActorIdentity` message if no actor could be found for the `Identify` message. I assume that this is the case if Akka cannot reach the remote actor system or if the remote actor system tells that there is no such actor with the given Akka URL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/registration/RetryingRegistration.java (hunk 1 above)
Comment (StephanEwen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/registration/RetryingRegistration.java]: Ah, cool, did not know that. Good then...

**Comment by StephanEwen** [comment]
Makes sense. +1 to merge this.

**Comment by tillrohrmann** [comment]
Failing test cases are unrelated. Will be merging this PR.

**Pull request closed without merging**